zulip/client.go

```diff
@@ -0,0 +1,182 @@
+package zulip
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+)
+
+var defaultBaseURL *url.URL = must(url.Parse("https://recurse.zulipchat.com/api/v1/"))
+
+// Credentials is used to authenticate requests to the Zulip API.
+type Credentials struct {
+	Username string
+	Password string
+}
+
+// CredentialsFunc returns the current Zulip API credentials.
+type CredentialsFunc func(context.Context) (Credentials, error)
+
+type Client struct {
+	http        *http.Client
+	baseURL     *url.URL
+	credentials CredentialsFunc
+}
+
+// NewClient creates a new Zulip API client.
+func NewClient(credentials CredentialsFunc, opts ...ClientOpt) (*Client, error) {
+	client := Client{
+		http:        http.DefaultClient,
+		baseURL:     defaultBaseURL,
+		credentials: nil,
+	}
+
+	for i, opt := range opts {
+		if err := opt(&client); err != nil {
+			return nil, fmt.Errorf("zulip client option %d: %w", i, err)
+		}
+	}
+
+	return &client, nil
+}
+
+func (c *Client) postForm(ctx context.Context, endpoint *url.URL, form url.Values) error {
+	req, err := http.NewRequestWithContext(
+		ctx,
+		http.MethodPost,
+		endpoint.String(),
+		strings.NewReader(form.Encode()),
+	)
+	if err != nil {
+		return fmt.Errorf("build request: %w", err)
+	}
+
+	req.Header.Set("content-type", "application/x-www-form-urlencoded")
+
+	creds, err := c.credentials(ctx)
+	if err != nil {
+		return fmt.Errorf("fetch credentials: %w", err)
+	}
+	req.SetBasicAuth(creds.Username, creds.Password)
+
+	resp, err := c.http.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	// This read will consume the body...
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("read response body: %w", err)
+	}
+
+	// ... so replace the content afterward.
+	resp.Body = io.NopCloser(bytes.NewReader(body))
+
+	log.Printf("zulip response: %d %s\n", resp.StatusCode, string(body))
+
+	if resp.StatusCode >= 400 {
+		return &ResponseError{resp}
+	}
+	return nil
+}
+
+// PostToTopic sends a chat message to the given stream and topic.
+func (c *Client) PostToTopic(ctx context.Context, stream, topic, message string) error {
+	// TODO: Stub out in non-prod, which should not send publicly-visible messages.
```

Comment:
Oh yes, something I skipped while porting. Teach the client about the app env for now.

Maybe consider sending to one of the bot-test streams instead?